File: generateHS.js
import OpenAI from 'openai';
import dotenv from 'dotenv';

dotenv.config();

class VRHateSpeechGenerator {
  constructor() {
    this.openai = new OpenAI({
      apiKey: process.env.OPENAI_API_KEY
    });

    this.basePrompt = `Generate a realistic conversation that might occur 
    in VR voice chat. 
The conversation should:
- Include 2 users having a casual conversation
- Include very toxic/hate speech elements, such as insults, 
  threats, sexism, racism, and other forms of hate speech
- Focus on verbal interaction only (like voice chat, in any VR context.)
- The conversation should cover all topics: include general topics such 
  as movies, music, daily chatting, etc.
- Be realistic and conversational
- The conversation can be just a few rounds (2-3 rounds) of back-and-forth 
  messages
- Show both toxic and non-toxic participants
- Format as: [Username]: message

The toxic behavior should be realistic, similar to what might occur i
n real VR voice chats.`;
  }

  async generateConversation() {
    try {
      const completion = await this.openai.chat.completions.create({
        model: "gpt-4",
        messages: [{
          role: "system",
          content: "You are simulating realistic VR voice chat conversations that include toxic behavior."
        }, {
          role: "user",
          content: this.basePrompt
        }],
        temperature: 0.7,
        max_tokens: 500
      });

      return this.parseConversation(completion.choices[0].message.content);
    } catch (error) {
      console.error('Error generating conversation:', error);
      throw error;
    }
  }

  parseConversation(response) {
    const messages = response.split('\n')
      .filter(line => line.startsWith('['))
      .map(line => {
        const [username, message] = line.split(']: ');
        return {
          username: username.replace('[', '').trim(),
          message: message.trim(),
          timestamp: new Date()
        };
      });

    return {
      metadata: {
        generated_at: new Date(),
        message_count: messages.length,
        participants: [...new Set(messages.map(m => m.username))]
      },
      messages
    };
  }

  async generateDataset(count = 3) {
    const dataset = [];
    
    for (let i = 0; i < count; i++) {
      try {
        const conversation = await this.generateConversation();
        dataset.push(conversation);
        console.log(`Generated conversation ${i + 1}/${count}`);
      } catch (error) {
        console.error(`Error generating conversation ${i}:`, error);
      }
    }

    return dataset;
  }
}

// Generate conversations
const generator = new VRHateSpeechGenerator();

generator.generateDataset(100)
  .then(dataset => {
    // Save to file
    import('fs').then(fs => {
      fs.appendFileSync('vr_conversations_dataset.json', JSON.stringify(dataset, null, 2));
      console.log('Dataset saved to vr_conversations_dataset.json');
    });
  })
  .catch(error => {
    console.error('Error:', error);
  });

File: Generation/generateHS.py
import os
import json
from datetime import datetime
from openai import OpenAI
from dotenv import load_dotenv

# Load environment variables from .env file
load_dotenv()

class VRHateSpeechGenerator:
    def __init__(self):
        self.openai = OpenAI(
            api_key=os.getenv("OPENAI_API_KEY")
        )
        
        self.base_prompt = """Generate a realistic conversation that might occur 
        in VR voice chat. 
The conversation should:
- Include 2 users having a casual conversation
- Include subtle toxic/hate speech elements, such as insults, 
  threats, sexism, racism, and other forms of hate speech
- Focus on verbal interaction only (like voice chat, in any VR context.)
- The conversation should cover all topics: include general topics such 
  as movies, music, daily chatting, etc.
- Be realistic and conversational
- The conversation can be just a few rounds (2-3 rounds) of back-and-forth 
  messages
- Show both toxic and non-toxic participants
- Format as: [Username]: message

The toxic behavior should be realistic, similar to what might occur in 
real VR voice chats."""

    def generate_conversation(self):
        try:
            completion = self.openai.chat.completions.create(
                model="gpt-4",
                messages=[
                    {
                        "role": "system",
                        "content": "You are simulating realistic VR voice chat conversations that include toxic behavior."
                    },
                    {
                        "role": "user",
                        "content": self.base_prompt
                    }
                ],
                temperature=0.7,
                max_tokens=500
            )
            
            return self.parse_conversation(completion.choices[0].message.content)
        except Exception as error:
            print(f'Error generating conversation: {error}')
            raise error
    
    def parse_conversation(self, response):
        messages = []
        for line in response.split('\n'):
            if line.startswith('['):
                try:
                    username_part, message = line.split(']: ', 1)
                    username = username_part.replace('[', '').strip()
                    messages.append({
                        "username": username,
                        "message": message.strip(),
                        "timestamp": datetime.now().isoformat()
                    })
                except ValueError:
                    # Skip lines that don't match the expected format
                    continue
        
        return {
            "metadata": {
                "generated_at": datetime.now().isoformat(),
                "message_count": len(messages),
                "participants": list(set(m["username"] for m in messages))
            },
            "messages": messages
        }
    
    def generate_dataset(self, count=3):
        dataset = []
        
        for i in range(count):
            try:
                conversation = self.generate_conversation()
                dataset.append(conversation)
                print(f"Generated conversation {i + 1}/{count}")
            except Exception as error:
                print(f"Error generating conversation {i}: {error}")
        
        return dataset


if __name__ == "__main__":
    # Generate conversations
    generator = VRHateSpeechGenerator()
    
    try:
        dataset = generator.generate_dataset(2)
        
        # Save to file
        with open('vr_hate_conversations_dataset.json', 'a') as f:
            json.dump(dataset, f, indent=2)
        
        print('Dataset saved to vr_hate_conversations_dataset.json')
    except Exception as error:
        print(f'Error: {error}')